pipeline/alignments/align.py

```diff
@@ -153,17 +192,105 @@ def write_priors(
     fwd_path: str,
     rev_path: str,
     priors_output_path: str,
-    stack: ExitStack,
 ):
+    import eflomal
+
     logger.info("Calculating priors...")
-    src_input = stack.enter_context(open(corpus_src, "r", encoding="utf-8"))
-    trg_input = stack.enter_context(open(corpus_trg, "r", encoding="utf-8"))
-    fwd_f = stack.enter_context(open(fwd_path, "r", encoding="utf-8"))
-    rev_f = stack.enter_context(open(rev_path, "r", encoding="utf-8"))
-    priors_tuple = eflomal.calculate_priors(src_input, trg_input, fwd_f, rev_f)
-    logger.info(f"Writing priors to {priors_output_path}...")
-    priors_output = stack.enter_context(open(priors_output_path, "w", encoding="utf-8"))
-    eflomal.write_priors(priors_output, *priors_tuple)
+    with ExitStack() as stack:
+        src_input = stack.enter_context(open(corpus_src, "r", encoding="utf-8"))
+        trg_input = stack.enter_context(open(corpus_trg, "r", encoding="utf-8"))
+        fwd_f = stack.enter_context(open(fwd_path, "r", encoding="utf-8"))
+        rev_f = stack.enter_context(open(rev_path, "r", encoding="utf-8"))
+        priors_tuple = eflomal.calculate_priors(src_input, trg_input, fwd_f, rev_f)
+        logger.info(f"Writing priors to {priors_output_path}...")
+        priors_output = stack.enter_context(open(priors_output_path, "w", encoding="utf-8"))
+        eflomal.write_priors(priors_output, *priors_tuple)
+
+
+def remap(
+    src_path: str,
+    trg_path: str,
+    tok_src_path: str,
+    tok_trg_path: str,
+    aln_path: str,
+    output_aln_path: str,
+) -> None:
+    """
+    Remaps alignments that were calculated for Moses-tokenized corpus to whitespace-tokenized ones.
+    :param src_path: path to whitespace-tokenized sentences in source language
+    :param trg_path: path to whitespace-tokenized sentences in target language
+    :param tok_src_path: path to Moses-tokenized sentences in source language
+    :param tok_trg_path: path to Moses-tokenized sentences in target language
+    :param aln_path: path to the alignments calculated for Moses-tokenized corpus
+    :param output_aln_path: path to output alignments file remapped to whitespace-tokenized corpus
+    """
+    logger.info("Remapping alignments to whitespace tokenization")
+
+    with ExitStack() as stack:
+        pool = stack.enter_context(multiprocessing.Pool(processes=multiprocessing.cpu_count()))
+        # Buffering helps to minimize IO operations which speeds thing up significantly
+        output = stack.enter_context(open(output_aln_path, "w", buffering=500000))
+
+        lines = zip(
+            stack.enter_context(open(src_path)),
+            stack.enter_context(open(trg_path)),
+            stack.enter_context(open(tok_src_path)),
+            stack.enter_context(open(tok_trg_path)),
+            stack.enter_context(open(aln_path)),
+        )
+
+        # send lines to worker processes in chunks
+        for aln in tqdm(pool.imap(remap_line, lines, chunksize=10000), mininterval=10):
```

Comment:
Question: Did you profile any of these changes to see the impact? Parallelism can speed up tasks, but will increase the memory pressure. I'd be curious of the memory/timing differences here.

Reply:
I did not profile but I watched the CPU/memory utilization while running both locally (on 10M) and on a server (400M sentences). The bottleneck for tokenization and remapping is CPU, so I'm parallelizing it. There were no issues with memory at these stages. Memory utilization is a lot higher during the aligning process which happens inside eflomal and we don't have any control on it.